File: datastruct/tree/red-black-tree/src/RBTree.hs
-- RBTree.hs, Haskell Red-Black tree implementation
-- Copyright (C) 2010 Liu Xinyu (liuxinyu95@gmail.com)
--
-- This program is free software: you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation, either version 3 of the License, or
-- (at your option) any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this program.  If not, see <http://www.gnu.org/licenses/>.

-- Most of the work comes from Okasaki's work in [1]

module RBTree where

import Test.QuickCheck
import qualified Data.List as L -- for verification purpose only
import Prelude hiding (min)

data Color = R | B | BB deriving (Show, Eq) -- BB is doubly black for deletion
data RBTree a = Empty
              | Node Color (RBTree a) a (RBTree a)
              | BBEmpty -- doubly black empty

min (Node _ Empty x _) = x
min (Node _ l _ _) = min l

isEmpty Empty = True
isEmpty _ = False

insert x = makeBlack . ins x where
    ins x Empty = Node R Empty x Empty
    ins x (Node color l k r)
        | x < k     = balance color (ins x l) k r
        | otherwise = balance color l k (ins x r)
    makeBlack (Node _ l k r) = Node B l k r

balance B (Node R (Node R a x b) y c) z d = Node R (Node B a x b) y (Node B c z d) -- case 1
balance B (Node R a x (Node R b y c)) z d = Node R (Node B a x b) y (Node B c z d) -- case 2
balance B a x (Node R b y (Node R c z d)) = Node R (Node B a x b) y (Node B c z d) -- case 3
balance B a x (Node R (Node R b y c) z d) = Node R (Node B a x b) y (Node B c z d) -- case 4
balance color l k r = Node color l k r

delete x = makeBlack . del x where
    del x Empty = Empty
    del x (Node color l k r)
        | x < k = fixDB color (del x l) k r
        | x > k = fixDB color l k (del x r)
        -- x == k, delete this node
        | isEmpty l = if color == B then shiftBlack r else r
        | isEmpty r = if color == B then shiftBlack l else l
        | otherwise = fixDB color l m (del m r) where m = min r
    makeBlack (Node _ l k r) = Node B l k r
    makeBlack _ = Empty

shiftBlack (Node B l k r) = Node BB l k r
shiftBlack (Node _ l k r) = Node B  l k r
shiftBlack Empty = BBEmpty
shiftBlack BBEmpty = Empty

isDB (Node BB _ _ _) = True
isDB BBEmpty = True
isDB _ = False

-- To solve the uniform black height violation.
fixDB::Color -> RBTree a -> a -> RBTree a -> RBTree a
-- Case 1: The sibling of the doubly-black nod is black, and it has a red sub-tree.
fixDB color a@(Node BB _ _ _) x (Node B (Node R b y c) z d) = Node color (Node B (shiftBlack a) x b) y (Node B c z d)
fixDB color BBEmpty x (Node B (Node R b y c) z d) = Node color (Node B Empty x b) y (Node B c z d)
fixDB color a@(Node BB _ _ _) x (Node B b y (Node R c z d)) = Node color (Node B (shiftBlack a) x b) y (Node B c z d)
fixDB color BBEmpty x (Node B b y (Node R c z d)) = Node color (Node B Empty x b) y (Node B c z d)
fixDB color (Node B a x (Node R b y c)) z d@(Node BB _ _ _) = Node color (Node B a x b) y (Node B c z (shiftBlack d))
fixDB color (Node B a x (Node R b y c)) z BBEmpty = Node color (Node B a x b) y (Node B c z Empty)
fixDB color (Node B (Node R a x b) y c) z d@(Node BB _ _ _) = Node color (Node B a x b) y (Node B c z (shiftBlack d))
fixDB color (Node B (Node R a x b) y c) z BBEmpty = Node color (Node B a x b) y (Node B c z Empty)
-- Case 2: The sibling of the doubly-black node is red
fixDB B a@(Node BB _ _ _) x (Node R b y c) = fixDB B (fixDB R a x b) y c
fixDB B a@BBEmpty x (Node R b y c) = fixDB B (fixDB R a x b) y c
fixDB B (Node R a x b) y c@(Node BB _ _ _) = fixDB B a x (fixDB R b y c)
fixDB B (Node R a x b) y c@BBEmpty = fixDB B a x (fixDB R b y c)
-- Case 3: The sibling of the doubly-black node, and its two sub-trees are all black.
--         Propagate the blackness up.
fixDB color a@(Node BB _ _ _) x (Node B b y c) = shiftBlack (Node color (shiftBlack a) x (Node R b y c))
fixDB color BBEmpty x (Node B b y c) = shiftBlack (Node color Empty x (Node R b y c))
fixDB color (Node B a x b) y c@(Node BB _ _ _) = shiftBlack (Node color (Node R a x b) y (shiftBlack c))
fixDB color (Node B a x b) y BBEmpty = shiftBlack (Node color (Node R a x b) y Empty)
-- otherwise
fixDB color l k r = Node color l k r

-- Auxiliary functions

fromList::(Ord a)=>[a] -> RBTree a
fromList = foldr insert Empty

toList :: (Ord a) => RBTree a -> [a]
toList Empty = []
toList (Node _ l x r) = toList l ++ [x] ++ toList r

-- For pretty printing
instance Show a => Show (RBTree a) where
    show BBEmpty = "*"
    show Empty = "."
    show (Node c l k r) = "(" ++ show l ++ " " ++
                          show k ++ ":" ++ show c ++ " " ++
                          show r ++ ")"

-- test

isRootBlack :: (RBTree a) -> Bool
isRootBlack Empty = True
isRootBlack (Node B _ _ _) = True
isRootBlack _ = False

adjacentRed :: (RBTree a) -> Bool
adjacentRed Empty = False
adjacentRed (Node R l@(Node R _ _ _) _ _) = True
adjacentRed (Node R _ _ r@(Node R _ _ _)) = True
adjacentRed (Node _ l _ r) = adjacentRed l || adjacentRed r

eqBlack t = blackness t > 0

blackness Empty = 1
blackness (Node c l _ r) = if a /= b then -1000 else (a + (ord c))
  where
    a = blackness l
    b = blackness r
    ord B = 1
    ord _ = 0

isRedBlack :: (RBTree a) -> Bool
isRedBlack t = and [isRootBlack t, not $ adjacentRed t, eqBlack t]

prop_bst :: (Ord a, Num a) => [a] -> Bool
prop_bst xs = let xs' = L.nub xs in
  (L.sort xs') == (toList $ fromList xs')

prop_insert_redblack :: (Ord a, Num a) => [a] -> Bool
prop_insert_redblack = isRedBlack . fromList . L.nub

prop_del :: (Ord a, Num a) => [a] -> a -> Bool
prop_del xs x = let xs' = L.nub xs in
  L.sort (L.delete x xs') == toList (delete x (fromList xs'))

prop_del_redblack :: (Ord a, Num a) => [a] -> a -> Bool
prop_del_redblack xs x = let xs' = L.nub xs in
  isRedBlack $ delete x (fromList xs')

testAll = do
  quickCheck (prop_bst::[Int]->Bool)
  quickCheck (prop_insert_redblack::[Int]->Bool)
  quickCheck (prop_del::[Int]->Int->Bool)
  quickCheck (prop_del_redblack::[Int]->Int->Bool)

{- Reference
[Chris Okasaki] Functional Pearls Red-Black trees in a functional setting.
   J. Functional programming. Jan. 1999.
   http://www.eecs.usma.edu/webs/people/okasaki/jfp99.ps-}

File: datastruct/tree/red-black-tree/src/scala/RBTree.scala
/*
 * RBTree.scala
 * Copyright (C) 2018 Liu Xinyu (liuxinyu95@gmail.com)
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 * 
 */
import scala.util.Random   //for verification
import scala.language.postfixOps

object RBTree {
  sealed abstract class Color
  case object R extends Color
  case object B extends Color
  case object BB extends Color  //doubly black

  sealed trait Tree[+A]
  case object E extends Tree[Nothing]
  case object BBE extends Tree[Nothing] //doubly black empty
  case class T[A] (c: Color, left: Tree[A], key: A, right: Tree[A]) extends Tree[A]

  def insert[A <% Ordered[A]] (tr: Tree[A], x: A): Tree[A] = makeBlack(ins(tr, x))

  def ins[A <% Ordered[A]] (tr: Tree[A], x: A): Tree[A] = tr match {
    case E => T(R, E, x, E)
    case T(c, left, y, right) =>
      if (x < y) balance(c, ins(left, x), y, right)
      else balance(c, left, y, ins(right, x))
    case _ => throw new RuntimeException("shouldn't be here")
  }

  def makeBlack[A](tr: Tree[A]): Tree[A] = tr match {
    case T(_, l, k, r) => T(B, l, k, r)
    case _ => E
  }

  def balance[A](c: Color, left: Tree[A], key: A, right: Tree[A]) : Tree[A] =
    (c, left, key, right) match {
      case (B, T(R, T(R, a, x, b), y, c), z, d) => T(R, T(B, a, x, b), y, T(B, c, z, d))
      case (B, T(R, a, x, T(R, b, y, c)), z, d) => T(R, T(B, a, x, b), y, T(B, c, z, d))
      case (B, a, x, T(R, b, y, T(R, c, z, d))) => T(R, T(B, a, x, b), y, T(B, c, z, d))
      case (B, a, x, T(R, T(R, b, y, c), z, d)) => T(R, T(B, a, x, b), y, T(B, c, z, d))
      case _ => T(c, left, key, right)
    }

  def delete[A <% Ordered[A]](tr: Tree[A], x: A): Tree[A] = makeBlack(del(tr, x))

  def del[A <% Ordered[A]](tr: Tree[A], x: A): Tree[A] = tr match {
    case E => E
    case T(color, l, k, r) =>
      if (x < k)
        fixDel(color, del(l, x), k, r)
      else if (x > k)
        fixDel(color, l, k, del(r, x))
      else if (isEmpty(l))
        color match {
          case B => addBlack(r)
          case _ => r
        }
      else if (isEmpty(r))
        color match {
          case B => addBlack(l)
          case _ => l
        }
      else {
        val k1 = min(r)
        fixDel(color, l, k1, del(r, k1))
      }
    case _ => throw new RuntimeException("shouldn't be here")
  }

  def isEmpty[A](tr: Tree[A]): Boolean = tr match {
    case E => true
    case _ => false
  }

  def addBlack[A](tr: Tree[A]): Tree[A] = tr match {
    case T(B, l, k, r) => T(BB, l, k, r)
    case T(_, l, k, r) => T(B, l, k, r)
    case E => BBE
    case _ => tr
  }

  def min[A] (tr: Tree[A]): A = tr match {
    case T(_, E, x, _) => x
    case T(_, left, _, _) => min(left)
  }

  def fixDel[A] (color: Color, left: Tree[A], key: A, right: Tree[A]) : Tree[A] =
    (color, left, key, right) match {
      // the sibling is black, and has one red nephew
      case (color, T(BB, _, _, _), x, T(B, T(R, b, y, c), z, d))
        => T(color, T(B, addBlack(left), x, b), y, T(B, c, z, d))
      case (color, BBE, x, T(B, T(R, b , y, c), z, d))
        => T(color, T(B, E, x, b), y, T(B, c, z, d))
      case (color, T(BB, _, _, _), x, T(B, b, y, T(R, c, z, d)))
        => T(color, T(B, addBlack(left), x, b), y, T(B, c, z, d))
      case (color, BBE, x, T(B, b, y, T(R, c, z, d)))
        => T(color, T(B, E, x, b), y, T(B, c, z, d))
      case (color, T(B, a, x, T(R, b, y, c)), z, T(BB, _, _, _))
        => T(color, T(B, a, x, b), y, T(B, c, z, addBlack(right)))
      case (color, T(B, a, x, T(R, b, y, c)), z, BBE)
        => T(color, T(B, a, x, b), y, T(B, c, z, E))
      case (color, T(B, T(R, a, x, b), y, c), z, T(BB, _, _, _))
        => T(color, T(B, a, x, b), y, T(B, c, z, addBlack(right)))
      case (color, T(B, T(R, a, x, b), y, c), z, BBE)
        => T(color, T(B, a, x, b), y, T(B, c, z, E))
      // the sibling is red
      case (B, T(BB, _, _, _), x, T(R, b, y, c))
        => fixDel(B, fixDel(R, left, x, b), y, c)
      case (B, BBE, x, T(R, b, y, c))
        => fixDel(B, fixDel(R, BBE, x, b), y, c)
      case (B, T(R, a, x, b), y, T(BB, _, _, _))
        => fixDel(B, a, x, fixDel(R, b, y, right))
      case (B, T(R, a, x, b), y, BBE)
        => fixDel(B, a, x, fixDel(R, b, y, BBE))
      // the slibing and 2 newphews are all black, propagate the blackness up
      case (color, T(BB, _, _, _), x, T(B, b, y, c))
        => addBlack(T(color, addBlack(left), x, T(R, b, y, c)))
      case (color, BBE, x, T(B, b, y, c))
        => addBlack(T(color, E, x, T(R, b, y, c)))
      case (color, T(B, a, x, b), y, T(BB, _, _, _))
        => addBlack(T(color, T(R, a, x, b), y, addBlack(right)))
      case (color, T(B, a, x, b), y, BBE)
        => addBlack(T(color, T(R, a, x, b), y, E))
      // otherwise
      case _ => T(color, left, key, right)
    }

  def fromList [A <% Ordered[A]] (xs: Seq[A]) : Tree[A] = ((E: Tree[A]) /: xs) (insert)

  def toList[A] (tr: Tree[A]) : List[A] = tr match {
    case E => List()
    case T(_, left, x, right) => toList(left) ::: (x :: toList(right))
    case _ => throw new RuntimeException("shouldn't be here")
  }

  // test
  def isBlack[A](tr: Tree[A]) : Boolean = tr match {
    case E => true
    case T(B, _, _, _) => true
    case _ => false
  }

  def adjacentRed[A](tr: Tree[A]) : Boolean = tr match {
    case E => false
    case T(R, T(R, _, _, _), _, _) => true
    case T(R, _, _, T(R, _, _, _)) => true
    case T(_, left, _, right) => adjacentRed(left) || adjacentRed(right)
  }

  def eqBlack[A](tr: Tree[A]) = blackness(tr) > 0

  def blackness[A](tr: Tree[A]) : Int = tr match {
    case E => 1
    case T(c, left, _, right) => {
      val a = blackness(left)
      val b = blackness(right)
      val i = if (isBlack(tr)) 1 else 0
      if (a != b) -1000 else a + i
    }
    case _ => throw new RuntimeException("shouldn't be here")
  }

  def isRedBlack[A](tr: Tree[A]) = isBlack(tr) && (!adjacentRed(tr)) && eqBlack(tr)

  val N = 100
  def genList(r: Random) = r.shuffle(0 to N - 1).take(r.nextInt(N))

  def testBuild(xs: Seq[Int]) = {
    val tr = fromList(xs)
    assert(toList(tr) == xs.sortWith(_ < _), println("violate build invariant"))
    assert(isRedBlack(tr), println("violate red-black properties"))
  }

  def testDelete(xs: Seq[Int]) = {
    if (!xs.isEmpty) {
      xs.foldLeft((xs.sorted, fromList(xs))) {
        (t, x) => {
          val ys = t._1
          val tr = t._2
          assert(ys == toList(tr), println(s"inconsist delete result: $ys"))
          assert(isRedBlack(tr), println("violate red-black properties"))
          (ys diff List(x), delete(tr, x))
        }
      }
    }
  }

  def test() = {
    val r = Random
    for (_ <- 1 to N) {
      val xs = genList(r)
      testBuild(xs)
      testDelete(xs)
    }
    println(s"$N tests passed");
  }
}
